### PR TITLE
Align GrpcSender contract with HttpSender

### PR DESCRIPTION
**Pull request by jack-berg**
I was reviewing #6645 I was cringing at inconsistencies between GrpcSender and HttpSender. 

- GrpcSender has `void send(T request, Runnable onSuccess, BiConsumer<GrpcResponse, Throwable> onError)`
- HttpSender has `void send(Marshaler marshaler, int contentLength, Consumer<Response> onResponse, Consumer<Throwable> onError);`

HttpSender got it right by cleanly separating out the difference between an export request that received a response, and one that ended exceptionally. Its up to the caller HttpExporter to evaluate the response and decide whether it was a success or error response. For GrpcSender, these are muddied together in a `BiConsumer<GrpcResponse, Throwable>`.

For #6645, when export attempts fail, we want to include exception in the `CompletableResultCode` which allow the caller to interpret what happened. Cleanly delineating between when a export failed but received a response, and failed exceptionally makes for a much cleaner API and user experience.

### CODE REVIEW

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcSender.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
  */
 public interface GrpcSender<T extends Marshaler> {
 
-  void send(T request, Runnable onSuccess, BiConsumer<GrpcResponse, Throwable> onError);
+  void send(T request, Consumer<GrpcResponse> onResponse, Consumer<Throwable> onError);
```
Comment (breedx-splk) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcSender.java:20]: Yeah, this is a good catch.

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java [hunk 2]
```diff
@@ -62,23 +61,59 @@ public CompletableResultCode export(T exportRequest, int numItems) {
...
+            exporterMetrics.addSuccess(numItems);
+            result.succeed();
+            return;
+          }
 
-          switch (response.grpcStatusValue()) {
-            case GRPC_STATUS_UNKNOWN:
-              result.failExceptionally(FailedExportException.grpcFailedExceptionally(throwable));
+          exporterMetrics.addFailed(numItems);
+          switch (grpcResponse.grpcStatusValue()) {
```
Comment (breedx-splk) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java:74]:
```suggestion
          switch (statusCode) {
```

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java [hunk 3]
```diff
@@ -62,23 +61,59 @@ public CompletableResultCode export(T exportRequest, int numItems) {
 
     grpcSender.send(
         exportRequest,
-        () -> {
-          exporterMetrics.addSuccess(numItems);
-          result.succeed();
-        },
-        (response, throwable) -> {
-          exporterMetrics.addFailed(numItems);
+        grpcResponse -> {
```
Comment (breedx-splk) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java:64]: I think the readability could be improved by making `onSuccess` and `onError` into functional instance methods.

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java [hunk 4]
```diff
@@ -62,23 +61,59 @@ public CompletableResultCode export(T exportRequest, int numItems) {
...
+                      + grpcResponse.grpcStatusDescription());
               break;
             default:
-              result.failExceptionally(FailedExportException.grpcFailedWithResponse(response));
+              logger.log(
+                  Level.WARNING,
+                  "Failed to export "
+                      + type
+                      + "s. Server responded with gRPC status code "
+                      + grpcResponse.grpcStatusValue()
```
Comment (breedx-splk) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java:97]:
```suggestion
                      + statusCode
```

File: exporters/sender/grpc-managed-channel/src/main/java/io/opentelemetry/exporter/sender/grpc/managedchannel/internal/UpstreamGrpcSender.java [hunk 5]
```diff
@@ -71,19 +73,41 @@ public void send(T request, Runnable onSuccess, BiConsumer<GrpcResponse, Throwab
...
+                GrpcResponse.create(Status.OK.getCode().value(), Status.OK.getDescription()));
           }
 
           @Override
           public void onFailure(Throwable t) {
-            Status status = Status.fromThrowable(t);
-            onError.accept(
-                GrpcResponse.create(status.getCode().value(), status.getDescription()), t);
+            Status status = fromThrowable(t);
+            if (status != null) {
```
Comment (breedx-splk) [inline exporters/sender/grpc-managed-channel/src/main/java/io/opentelemetry/exporter/sender/grpc/managedchannel/internal/UpstreamGrpcSender.java:83]: Nit: Prefer to invert this if.

**Review by breedx-splk — APPROVED** [review]
Looks good, I like the alignment. I had a few small suggestions, but nothing to block this. Thanks!

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java (hunk 3 above)
Comment (jack-berg) [reply exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java:64]: Also applied to the same structure in HttpExporter

**Pull request merged**